rats-apps/test/python/rats_test/apps/test_annotated_container.py

```diff
@@ -0,0 +1,288 @@
+            account="foo-miss",
+            table="bar",
+            max=100,
+        )
+
+    @apps.fallback_config(ExampleIds.make_config_id("foo-config.miss-fallback"))
+    def fallback_foo_config_miss(self) -> ExampleConfig:
+        return ExampleConfig(
+            account="foo-miss-fallback",
+            table="bar",
+            max=100,
+        )
+
+    @apps.config(ExampleIds.make_config_id("foo-config.duplicate"))
+    def foo_config_duplicate_1(self) -> ExampleConfig:
+        return ExampleConfig(
+            account="foo-hit",
+            table="bar",
+            max=100,
+        )
+
+    @apps.config(ExampleIds.make_config_id("foo-config.duplicate"))
+    def foo_config_duplicate_2(self) -> ExampleConfig:
+        return ExampleConfig(
+            account="foo-hit",
+            table="bar",
+            max=100,
+        )
+
+
+class TestAnnotatedContainer:
+    _app: ExampleContainer
+
+    def setup_method(self) -> None:
+        self._app = ExampleContainer()
+
+    def test_services_api(self) -> None:
+        """
+        Basic API for retrieving services from annotated containers.
+
+        - you get a service by providing a ServiceId
+        - the ServiceId generic holds the type returned by .get()
+        - multiple calls to .get() return the same instance of a given ServiceId
+        - method verifies that exactly one service matches a given ServiceId
+        - if the service id is not found, we look for and return a fallback service
+        - if the fallback service is not found, we raise a ServiceNotFoundError
+        - the fallback service is ignored if the main service is found
+        - there must still be at most one match for the service within the fallback id
+        """
+        storage = self._app.get(ExampleIds.STORAGE)
+        assert storage.called == 0
+        storage.save("hello")
+        assert storage.called == 1
+
+        storage_2 = self._app.get(ExampleIds.STORAGE)
+        assert storage == storage_2
+        assert storage.called == storage_2.called
+
+        with pytest.raises(apps.ServiceNotFoundError):
+            self._app.get(ExampleIds.MISSING)
+
+        with pytest.raises(apps.DuplicateServiceError):
+            self._app.get(ExampleIds.DUPLICATE)
+
+        fallback = self._app.get(ExampleIds.EXECUTABLE)
+        fallback.execute()
+        assert fallback.called == 1
+
+        foo = self._app.get(ExampleIds.make_service_id("foo"))
+        foo.execute()
+        assert foo.called == 1
+
+    def test_config_api(self) -> None:
+        """
+        Basic API for retrieving configs from annotated containers.
+
+        - you get a config by providing a ConfigId
+        - the ConfigId generic holds the type returned by .get_config()
+        - configs are bound to NamedTuple types and should strictly be data structures
+        - configs should be immutable and should not have side effects
+        - multiple calls to .get_config() return the same instance
```

Comment:
this is not necessary if configs are immutable and do not have side effects.

Reply:
that's true if you follow the practices our group has been developing, but we can't really prevent people from making configs that are not deterministic, like this:
```python
def my_config(self) -> NamedTuple:
  return DbSettings(
    host="localhost:8080",
    shard=randint(1, 10),
  )
```

i've encoded the fact that the same instance of `DbSettings` will get returned because i want to say that any dynamic aspects like the above need to be written into the library layer. so if your config has aspects that have to change over time, you need to create a service with a method that returns the proper config and we won't try to manage any kind of cache expiration logic

Reply:
got it